test/release/examples/benchmarks/shootout/fasta.chpl

```diff
@@ -96,47 +101,93 @@ proc repeatMake(desc, alu, n) {
 }
 
 //
-// Output a random sequence of length 'n' using distribution a
+// Output a random sequence of length 'n' using distribution 'a'
 //
-proc randomMake(desc, nucleotides, n) {
+proc randomMake(desc, nuclInfo, n) {
   stdout.writeln(desc);
 
-  const numNucls = nucleotides.size;
-  var cumulProb: [1..numNucls] int;
+  const numNucls = nuclInfo.size;
+  var cumulProb: [1..numNucls] randType;
+
+  // compute the cumulative probabilities of the nucleotides
   var p = 0.0;
   for i in 1..numNucls {
-    p += nucleotides[i](prob);
-    cumulProb[i] = 1 + (p*IM):int;
+    p += nuclInfo[i](prob);
+    cumulProb[i] = 1 + (p*IM):randType;
   }
 
-  var line_buff: [0..lineLength] int(8);
+  // guard when tasks can access the random numbers or output stream
+  var randGo, outGo: [0..#numTasks] atomic int;
+  randGo.write(1);
+  outGo.write(1);
+
+  coforall tid in 0..#numTasks {
+    const chunkSize = lineLength*blockSize,
+          nextTask = (tid + 1) % numTasks;
+
+    var myBuff: [0..(lineLength+1)*blockSize-1] int(8),
+        myRands: [0..chunkSize] int/*(32)*/;
+
+    for i in 1..n by chunkSize*numTasks align 1+tid*chunkSize {
+      const bytes = min(chunkSize, n-i+1);
+
+      // Get 'bytes' random numbers in a coordinated manner
+      wait(randGo);
+      getRands(bytes, myRands);
+      signal(randGo);
+
+      // Compute 'bytes' nucleotides and store in 'myBuff'
+      var col = 0,
+          off = 0;
+
+      for i in 0..#bytes {
+
+        const r = myRands[i];
+        var ncnt = 1;
+        for j in 1..numNucls do
+          if r >= cumulProb[j] then
+            ncnt += 1;
 
-  for i in 1..n by lineLength {
-    const bytes = min(lineLength, n-i+1);
+        myBuff[off] = nuclInfo[ncnt](nucl);
+        off += 1;
+        col += 1;
 
-    for (r, i) in zip(getRands(bytes), 0..) {
-      for j in 1..numNucls {
-        if r < cumulProb[j] {
-          line_buff[i] = nucleotides[j](nucl);
-          break;
+        if (col == lineLength) {
+          col = 0;
+          myBuff[off] = newline;
+          off += 1;
         }
       }
-    }
-    line_buff[bytes] = newline:int(8);
+      if (col != 0) {
+        myBuff[off] = newline;
+        off += 1;
+      }
+
+      // Write the output in a coordinated manner
+      wait(outGo);
+      stdout.write(myBuff[0..#off]);
+      signal(outGo);
+
+      // Helper routines for taking turns with shared resources
+      inline proc wait(guard) {
```

Comment:
It might be interesting to try out the ``waitFor`` method to see if performance is different, but this is fine by me for this PR.